source/opt/loop_descriptor.cpp

```diff
@@ -453,25 +453,59 @@ bool Loop::IsLCSSA() const {
   return true;
 }
 
-bool Loop::ShouldHoistInstruction(IRContext* context, Instruction* inst) {
-  return AreAllOperandsOutsideLoop(context, inst) &&
-         inst->IsOpcodeCodeMotionSafe();
+bool Loop::ShouldHoistInstruction(const Instruction& inst) const {
+  return inst.IsOpcodeCodeMotionSafe() && AreAllOperandsOutsideLoop(inst) &&
+         AreInOperandsInvariantInLoop(inst);
```

Comment:
Otherwise, there are lots of cases you will still have to consider.  You will probably want to do something like:

```
bool Loop::ShouldHoistInstruction(const Instruction& inst) const {
  return inst.IsOpcodeCodeMotionSafe() && AreAllOperandsOutsideLoop(inst) &&
         (!inst.IsLoad() || LoadedMemoryIsLoopInvariant(inst));
}
```

`LoadedMemoryIsLoopInvariant` is a new function that looks for stores that could potentially modify the memory that are in the loop. That is more than just OpStore. I can't find an existing function that will tell you if it does.  Even with that, you will have to watch out for memory barriers. Some loads cannot move past a memory barrier.

I recommend the simple fix.

Reply:
Great! That is a much simpler way to solve the issue. Thanks for looking into this.
I tested it with one of my projects, and every shader is now working as expected with LICMPass enabled.